Add support for custom Smelting-Recipes

TechnicianLP opened the pull request:
Alternative implementation to #5038 without an event
Adds a way to have the output/experience of a SmeltingRecipe to be dependant on the Input

I removed the now unneccessary patches from the FurnaceRecipes class

Test Mod modified from #5038; it is somewhat broken as potions (sometimes) get applied twice yet it shows the feature working

Slight binary incompatibility if someone were to directly use `FurnaceRecipes.instance().getSmeltingList().add()` to register their recipes as the now return map doesnt support the put operation (clear/remove should work though)

@mezz something else required for jei to work with this?

TechnicianLP committed:
Custom Smelting Recipes

TechnicianLP committed:
Add missing License Headers

TechnicianLP committed:
Fix recipes-field not being initialized

TechnicianLP committed:
Rerun genPatches

TechnicianLP committed:
Add TestMod

kashike commented:
This should be deprecated - 1.13 has support for custom smelting recipes.

TechnicianLP commented:
1.12 will have another year of support though and because of that i dont like the idea of putting `@Deprecated` on a new feature already

bs2609 commented:
Reminder to refer to #3450 (and previous #3141) if you haven't already.

liach commented on src/main/java/net/minecraftforge/common/smelting/SmeltingHandler.java:
```diff
@@ -0,0 +1,109 @@
+    {
+        if (recipes != null)
+            return recipes;
+        recipes = Maps.newHashMap();
```
I suggest using array-backed maps instead because the map is always iterated over (itemstack class does not override hashcode or equals)

liach replied:
By the way, you can just initialize this in the constructor. It won't bring too large an overhead while it makes code cleaner.

TechnicianLP replied:
using an array-backed map would probably make sense - as for being able to initialize it in the constructor: it isnt possible as the super-constructor calls overridden methods that require the field to be set

liach commented on src/main/java/net/minecraftforge/common/smelting/SmeltingHandler.java:
```diff
@@ -0,0 +1,109 @@
+    @Override
+    public Map<ItemStack, ItemStack> getSmeltingList()
+    {
+        return staticView;
```
The returned result is not safe here as it can be null

liach commented on src/main/java/net/minecraftforge/common/smelting/ISmeltingRecipe.java:
```diff
@@ -0,0 +1,36 @@
+
+public interface ISmeltingRecipe
+{
+    ItemStack getSmeltingResult(ItemStack input);
```
You should not in javadocs about the contract of this method that the returned item stack is not safely mutable (it should be copied before use). This is a behavior different from `IRecipe` that the crafting result of `IRecipe` is always copied before return, which can cause more confusions. Consider making a change to clarify this.

TechnicianLP replied:
forgot the copy

KnightMiner commented on patches/minecraft/net/minecraft/item/crafting/FurnaceRecipes.java.patch:
```diff
@@ -1,20 +1,11 @@
-@@ -113,6 +113,7 @@
+@@ -20,7 +20,7 @@
  
-     public void func_151394_a(ItemStack p_151394_1_, ItemStack p_151394_2_, float p_151394_3_)
```
Removing a vanilla function breaks binary compatibility

liach replied:
No, that is not removal. It is that old methods displayed in patches are no longer displayed because that piece of patch is now away.

KnightMiner replied:
Ah, right. Patches are a pain to read

KnightMiner commented on patches/minecraft/net/minecraft/item/crafting/FurnaceRecipes.java.patch:
```diff
@@ -1,20 +1,11 @@
      }
-@@ -142,6 +143,9 @@
  
-     public float func_151398_b(ItemStack p_151398_1_)
```
Binary compatibility

liach replied:
See my comment above.

KnightMiner commented:
I think using a list with a matches method is better than a map here as it would allow NBT sensitive inputs. With a map the output can still update based on NBT but you cannot match against a specific NBT of a recipe.

Also,  it might be nice to have support for cook time like #3450.

Also, I can assure you some modder is adding their furnace recipes in the way you removed, because moders do weird things. Since vanilla has a proper add method supporting XP, it might not be that big of a deal, but I'm not the best to answer that.

liach commented:
@KnightMiner The map is used here because vanilla also uses a map (although vanilla iterates through that map)

TechnicianLP committed:
Moved to Ingredient as Input
Changed CompoundIngredient to prevent theoretical SOE

TechnicianLP commented:
Changed the RecipeList to use ArrayList
Moved to Ingredient as Input (BUT: `Ingredient.getMatchingStacks()` cant match another recipes Ingredient (without having the same output; see merging below))
The VanillaMap view is now backed by a custom Map implementation that writes through to the underlying List
Implemeted basic Merging of Recipes if Inputs collide and outputs are the same: {A}>R + {A,B}>R = {A,A,B}>R (deduplication would require bigger changes to test for Ingrediant equals/contains)
